andrepereiradasilva/joomla-cms
# Detach language switcher from site app

### Summary of Changes

This PR aims to detach the language switcher from the `JApplicationSite` app.

To achieve this we also introduce a new method to `JApplicationCms`, the `setLanguage`method. This method allows to set a new app language while the page is being loaded.

Also we introduce a new user state `language` this user state allows to save the user langauge without the cookie (it uses the session cookie).

And many other code optimizations.
### Testing Instructions
1. Code review
2. Use latest staging with multilanguage (use several languages)
3. Apply patch
4. Test administrator language:
   - default admin language works fine
   - change default admin language in com_languages installed view works fine
   - switch admin language in com_languages installed view works fine
   - setting a user admin language in user profile/edit works fine
   - chosing a different admin language in the admin login works fine
5. Test frontend language with langauge switcher plugin DISABLED
   - default site language works fine
   - change default site language in com_languages installed view works fine
   - direct acess to language works fine `ex: index.php?language=fr-FR`
   - direct acess to language that does not exists fallback to en-GB fine `ex: index.php?language=xx-XX`
   - setting a user site language in user profile/edit works fine
6. Test frontend language with langauge switcher plugin ENABLED
   - default site language works fine
   - change default site language in com_languages installed view works fine
   - direct acess to language works fine `ex: index.php?language=fr-FR`
   - direct acess to language that does not exists fallback to en-GB fine `ex: index.php?language=xx-XX`
   - setting a user site language in user profile/edit works fine
   - check aceesing without language cookie fallback to defualt language.
7. Any other tests you could remeber
### Documentation Changes Required

None.
